egs/wsj/s5/steps/nnet3/make_cnn_tdnn_configs.py

```diff
@@ -0,0 +1,397 @@
+
+        cnn_reduced_ouput = args.cnn_reduced_dim;
+        print('component name=dim-reduce-affine type=NaturalGradientAffineComponent '
+              'input-dim={0} output-dim={1}'.format(maxp_output, cnn_reduced_ouput), file=f)
+
+    cur_dim = (nonlin_output_dim * len(splice_array[l-1]) if l > 1 else cnn_reduced_ouput + args.ivector_dim)
+
+    print('# Note: param-stddev in next component defaults to 1/sqrt(input-dim).', file=f)
+    print('component name=affine{0} type=NaturalGradientAffineComponent '
+          'input-dim={1} output-dim={2} bias-stddev=0'.
+        format(l, cur_dim, nonlin_input_dim), file=f)
+    if args.relu_dim is not None:
+        print('component name=nonlin{0} type=RectifiedLinearComponent dim={1}'.
+              format(l, args.relu_dim), file=f)
+    else:
+        print('# In nnet3 framework, p in P-norm is always 2.', file=f)
+        print('component name=nonlin{0} type=PnormComponent input-dim={1} output-dim={2}'.
+              format(l, args.pnorm_input_dim, args.pnorm_output_dim), file=f)
+    print('component name=renorm{0} type=NormalizeComponent dim={1}'.format(
+         l, nonlin_output_dim), file=f)
+    print('component name=final-affine type=NaturalGradientAffineComponent '
+          'input-dim={0} output-dim={1} param-stddev=0 bias-stddev=0'.format(
+          nonlin_output_dim, args.num_targets), file=f)
+    # printing out the next two, and their component-nodes, for l > 1 is not
+    # really necessary as they will already exist, but it doesn't hurt and makes
+    # the structure clearer.
+    if use_presoftmax_prior_scale :
+        print('component name=final-fixed-scale type=FixedScaleComponent '
+            'scales={0}/presoftmax_prior_scale.vec'.format(
+            args.config_dir), file=f)
+    print('component name=final-log-softmax type=LogSoftmaxComponent dim={0}'.format(
+          args.num_targets), file=f)
+    print('# Now for the network structure', file=f)
+    if l == 1:
+        splices = [ ('Offset(input, {0})'.format(n) if n != 0 else 'input') for n in splice_array[l-1] ]
+        #if args.ivector_dim > 0: splices.append('ReplaceIndex(ivector, t, 0)')
+        cur_input='Append({0})'.format(', '.join(splices))
+        # e.g. cur_input = 'Append(Offset(input, -2), ... Offset(input, 2), ivector)'
+        if args.use_mfcc == "true":    
+            print('component-node name=idct component=idct input={0}'.format("input"),
+                  file=f)
+            splices = [ ('Offset(idct, {0})'.format(n) if n != 0 else 'idct') for n in splice_array[l-1] ]
+            cur_input='Append({0})'.format(', '.join(splices))
+
+        if args.add_delta == "true":
+            print('component-node name={0}_conv component={0}_conv input={1}'.
+                  format("add-delta", cur_input), file=f)
+            cur_input="add-delta_conv"
+
+        for cl in range(0, num_cnn_layers):
+            print('component-node name=L{0}_conv component=L{0}_conv input={1}'.
```

Comment:
Please make use of the functions AddConvolutionNode() and AddMaxPoolNode() rather than writing this print statements. It is very difficult to follow.